common/txmgr/confirmer.go

```diff
@@ -294,178 +260,144 @@ func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Pro
+		}
+		// If unconfirmed transactions are identified as included, process them and mark them as confirmed or terminally stuck (if purge attempt exists)
+		err = ec.ProcessIncludedTxs(ctx, includedTxs, head)
+		if err != nil {
+			return fmt.Errorf("failed to process confirmed transactions: %w", err)
 		}
-
-		txIDsToUnconfirm = append(txIDsToUnconfirm, attempts[idx].TxID)
 	}
-	err = ec.txStore.UpdateTxsUnconfirmed(ctx, txIDsToUnconfirm)
 
-	if err != nil {
-		return err
-	}
-	return
+	return nil
 }
 
-// CheckForReceipts finds attempts that are still pending and checks to see if a receipt is present for the given block number.
-func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) CheckForReceipts(ctx context.Context, blockNum int64, latestFinalizedBlockNum int64) error {
-	attempts, err := ec.txStore.FindTxAttemptsRequiringReceiptFetch(ctx, ec.chainID)
-	if err != nil {
-		return fmt.Errorf("FindTxAttemptsRequiringReceiptFetch failed: %w", err)
-	}
-	if len(attempts) == 0 {
+func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) ProcessReorgTxs(ctx context.Context, reorgTxs []*txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], head types.Head[BLOCK_HASH]) error {
+	if len(reorgTxs) == 0 {
 		return nil
 	}
+	var etxIDs, attemptIDs []int64
+	for _, etx := range reorgTxs {
+		if len(etx.TxAttempts) == 0 {
+			return fmt.Errorf("invariant violation: expected tx %v to have at least one attempt", etx.ID)
+		}
 
-	ec.lggr.Debugw(fmt.Sprintf("Fetching receipts for %v transaction attempts", len(attempts)), "blockNum", blockNum)
-
-	attemptsByAddress := make(map[ADDR][]txmgrtypes.TxAttempt[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE])
-	for _, att := range attempts {
-		attemptsByAddress[att.Tx.FromAddress] = append(attemptsByAddress[att.Tx.FromAddress], att)
-	}
+		// Rebroadcast the one with the highest gas price
+		attempt := etx.TxAttempts[0]
 
-	for from, attempts := range attemptsByAddress {
-		minedSequence, err := ec.getMinedSequenceForAddress(ctx, from)
-		if err != nil {
-			return fmt.Errorf("unable to fetch pending sequence for address: %v: %w", from, err)
+		logValues := []interface{}{
+			"txhash", attempt.Hash.String(),
+			"currentBlockNum", head.BlockNumber(),
+			"currentBlockHash", head.BlockHash().String(),
+			"txID", etx.ID,
+			"attemptID", attempt.ID,
+			"nReceipts", len(attempt.Receipts),
+			"id", "confirmer",
 		}
 
-		// separateLikelyConfirmedAttempts is used as an optimisation: there is
-		// no point trying to fetch receipts for attempts with a sequence higher
-		// than the highest sequence the RPC node thinks it has seen
-		likelyConfirmed := ec.separateLikelyConfirmedAttempts(from, attempts, minedSequence)
-		likelyConfirmedCount := len(likelyConfirmed)
-		if likelyConfirmedCount > 0 {
-			likelyUnconfirmedCount := len(attempts) - likelyConfirmedCount
+		if attempt.State != txmgrtypes.TxAttemptBroadcast {
```

Comment:
What do we get from this check besides detection? Would there be a problem to rebroadcast this attempt?

Reply:
I don't think there would be an issue rebroadcasting this attempt. This validation carried over from the previous logic because we used to restrict the rebroadcast attempt to have to be broadcasted in `updateEthTxAttemptUnbroadcast`. But this helped me realize, we could potentially have in-progress attempts for transactions we mark as confirmed if the TXM goes down at the right time. I should mark any in-progress tx as broadcast when I mark tx as confirmed. Then I can also remove this check too.